// @ced/cli/lib/services/json-file-config.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const os = require("os");
const path = require("path");
const config_1 = require("../config/config");
class JsonFileConfigService {
    exists(file) {
        return fs.existsSync(file);
    }
    saveJson(file, obj) {
        const json = JSON.stringify(obj, null, 2);
        const dir = path.dirname(file);
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir);
        }
        fs.writeFileSync(file, json);
    }
    remove(file) {
        if (fs.existsSync(file)) {
            fs.unlinkSync(file);
        }
    }
    readJson(file) {
        const dir = path.dirname(file);
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir);
        }
        if (!fs.existsSync(file)) {
            return undefined;
        }
        const data = fs.readFileSync(file);
        const json = data.toString();
        return JSON.parse(json);
    }
    get globalConfigRootFolder() {
        const configFolder = config_1.config.environmentName === 'prod'
            ? '.ced'
            : '.ced-dev';
        return path.join(os.homedir(), configFolder);
    }
}
exports.JsonFileConfigService = JsonFileConfigService;
